### ISSUE TITLE
Support for PB741 (InkPad Color)

### ISSUE THREAD

**Issue by Popkornium18** [koreader/koreader#7406]
- KOReader version: v2021.02
- Device: PocketBook InkPad Color (PB741)

KOReader does not start on the device.

`crash.log`

```
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 12/31/10-11:02:48
 [*] Version: v2021.02

ffi.load: blitbuffer
ffi.load: SDL2
ffi.load (warning): libSDL2.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so
ffi.load (warning): libSDL2-2.0.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so.0
ffi.load (warning): libSDL2-2.0.so.0: cannot open shared object file: No such file or directory
SDL2 not loaded:	./setupkoenv.lua:27: Not able to load dynamic library: libSDL2-2.0.so.0
ffi.load: inkview
12/31/10-11:02:49 INFO  SoftwareVersion:  U741.6.3.560
./luajit: frontend/device/pocketbook/device.lua:607: unrecognized PocketBook model PB741
stack traceback:
	[C]: in function 'error'
	frontend/device/pocketbook/device.lua:607: in main chunk
	[C]: in function 'probeDevice'
	frontend/device.lua:53: in main chunk
	[C]: in function 'require'
	./reader.lua:144: in main chunk
	[C]: at 0x00013f31
!!!!
Uh oh, something went awry... (Crash n°1: 12/31/10 @ 11:02:49)
Attempting to restart KOReader . . .
!!!!
```

**Comment by duongkimngoc** [comment]
got the same thing here
please add support for PocketBook InkPad Color

**Comment by Frenzie** [comment]
Cf. https://github.com/koreader/koreader/pull/6680

It's hopefully more or less exactly the same.

**Comment by KnappeGEIL** [comment]
I've added the PB741 to device.lua and build a release for pocketbook. koreader starts and most things seem to be working but colors and wifi/networking don't. I could imagine that color depth needs to be set to 24bit somewhere for colors to work and from what I could gather wifi on pocketbook devices seems to be trash in general. Don't know why that is though.

PB741 FW: U741.6.3.560
koreader version: https://github.com/koreader/koreader/tree/b6b332e311a282f0f80dff24412197b2db5f1d6c

**Comment by NiLuJe** [comment]
You don't need to build anything to just tweak a few Lua files. And see the message right above yours for color.

**Comment by KnappeGEIL** [comment]
I've set `hasColorScreen = yes` and koreader tells me on startup that everything will be rendered in color and "Color rendering" is also enabled in the options but still everything is in greyscale. 

Also where could I start debugging the networking issues? I'm really new to koreader, lua and ebook readers in general.

**Comment by NiLuJe** [comment]
You need the koreader-base change, too.

**Comment by dustinjoe** [comment]
Hi, actually I just got my PB741 in hand and have not got Koreader working. Same crash problem...
Wondering if there is any solution update on this? Thanks.

**Comment by Frenzie** [comment]
That's why it's still open. ^_^

You'd need to apply these two to give it a try:
https://github.com/koreader/koreader-base/pull/1191
https://github.com/koreader/koreader/pull/6680

**Comment by dustinjoe** [comment]
Oh, thanks for your reply, will have a try on these.
Never player color eink devices before~

**Comment by NiLuJe** [comment]
Note that it *should* be supported (albeit in grayscale only) in current nightlies ;).

**Comment by Frenzie** [comment]
Current nightlies as in post-2021.05?

**Comment by NiLuJe** [comment]
Yup: https://github.com/koreader/koreader/commit/1ded7c3907eeb15f9dae780f17f7b090c00a7df2

**Comment by dustinjoe** [comment]
Thanks for the update.
Found the nightly builds here:
http://build.koreader.rocks/download/nightly/
Installed and working now. Thank you!

**Comment by Frenzie** [comment]
Yes, but a lot of it has already been done. See what I wrote in https://github.com/koreader/koreader/issues/7406#issuecomment-855876456

If someone with the device could try those changes and say if they work then we might be able to just merge it as is with a few minor adjustments, although I'll leave that decision up to @NiLuJe.

**Comment by Ingvix** [comment]
I've been trying to build it from the source to test it but right know luajit is failing due to incompatible /usr/lib64/gcc/x86_64-unknown-linux-gnu/10.2.1/libgcc.a. Help or maybe even just providing a prebuild package would be appreciated.

**Comment by Frenzie** [comment]
Since that's all in Lua the prebuilt package is the one you already have. ;-)

**Comment by Ingvix** [comment]
I meant prebuild package of KOReader with those changes because the only reason I'm pretty much trying to set this up is to try is to test them.

**Comment by Frenzie** [comment]
It's probably incompatible with https://github.com/koreader/koreader/pull/7415 in the full file form.

See https://github.com/koreader/koreader-base/pull/1191 and https://github.com/koreader/koreader/pull/6680 for the diffs.

**Comment by Ingvix** [comment]
Right, I'll see if I manage to patch them later today.

**Comment by NiLuJe** [comment]
@Frenzie: It's likely to be *extremely* slow without a C/NEON implementation :/.

**Comment by Frenzie** [comment]
Ah? I'd think the JIT code should be fairly speedy, possibly "slow" in more of a benchmark sense. You mean because of that weird reversal thing happening?

**Comment by dustinjoe** [comment]
Hi, I am playing around on my 741 for a while. One problem I am facing is that it does not seem to display contents(like color contents in pdf) in color. I tried to tap on the 'color rendering' in that menu but it does not seem to have effect. Is there any additional settings I need to have? Thanks.

**Comment by Frenzie** [comment]
That depends a bit on what you're talking about. If it's the nightly, that's indeed supposed to be grayscale.

If it's https://github.com/koreader/koreader-base/pull/1191 and https://github.com/koreader/koreader/pull/6680 then I don't know. I'm not sufficiently familiar with the relevant aspects of PB, or PB in general for that matter.

**Comment by Ingvix** [comment]
I managed to patch it earlier but saw no colors and also opening the file browser caused it to crash.

**Comment by debarupa** [comment]
Let me just start by saying how helpful this app is in PocketBook devices. I have both PocketBook Color and PocketBook Inkpad Color, and opening PDFs and page turns are considerably faster than the stock reading app .  So, kudos to everyone who made this possible. I'm using the latest build (Nov, 2021) for PocketBook. 

Any updates on supporting color yet? I'm experiencing the same problem as others here. On the startup screen, the koreader app says that it will render color, the same is checked in every book, but everything is displayed in grayscale.

**Comment by Frenzie** [comment]
See https://github.com/koreader/koreader/issues/7406#issuecomment-855876456

**Comment by rghe** [comment]
Hi,
I've tried both patches. Out of the box koreader crashes due to setter() requiring 7 parameters instead of five.
Fixed that, koreader works again, but without color.  
Strangely, only _few_ calls to blitDefault() happen with bpp>=24, triggering the CFA code.  
Most are done with bpp=8 (b/w ?) 
Perhaps the color mode is not really enabled ?

**Comment by poire-z** [comment]
(I may be off topic, so, just for info:
Most of our UI widgets are limited to 8bpp (except images we can show for book covers or Wikipedia results), so you'll get most blits (UI stuff) in 8bpp.
But the renderers can do colors (so, for book content in color), if `Gear> Screen> [x] Color rendering` is enabled. But they will draw on a buffer, and we'll just blit that on the screen framebuffer - so you may not see much logs/calls to these.
)

**Comment by rghe** [comment]
ok, I'll try with some plain color images

**Comment by debarupa** [comment]
Updated all 4 files per #7406. koreader starts up normally but when I try to go into directories, then it crashes. 

![20211124_193614](https://user-images.githubusercontent.com/2438991/143381247-90f578ae-6925-4856-875d-76a8716809ef.jpg)

**Comment by rghe** [comment]
koreader crashes because other changes affected the setter function, now requiring seven parameters instead of five.
The patches add these lines 

>         local function putPixel(x,y,value)
>            return setter(dest, x, y, value, set_param)
>        end

change the setter() call to

>         local function putPixel(x,y,value)
>            return setter(dest, x, y, value, set_param, x, y)
>        end

N.B.
koreader now works, but without color

**Comment by debarupa** [comment]
Thanks @rghe but koreader worked for me as is with the Nov 2021 build, but without color. I was trying these patches to see if it would show color but it didn't. I reverted the files back to the originals.

**Comment by NiLuJe** [comment]
Bugging PocketBook for a non-gimped sunxi driver and up-to-date kernels might be an approach, because we could just chuck that off to the kernel.

**Comment by rjd22** [comment]
I own a the pocketbook color and am thinking on spending some time to get color support working since it seems to be abandoned.

@ezdiy seems to have done a lot of awesome work on this but doesn't seem to be really active anymore so I am thinking of continuing his work.

Are there any big blockers on his work or implementation that kept it from being merged?

**Comment by Frenzie** [comment]
Not besides blitbuffer.c currently lacking in the base PR. It's mostly a matter of being able to try the changes on a few devices so as to not randomly break things.

**Comment by rjd22** [comment]
@Frenzie thanks, while I understand most of the changes I'm quite ignorant on what changes the blitsbuffer.c needs.

For the other PR I will remove the brute force enable so it's default off unless you have a color device.

I'm willing to do the work if someone could point me on what I need to add.

It would be great if it could end up in the nightly so it's easier for others to also test it.

**Comment by Frenzie** [comment]
The basis is just duplicating the work from the Lua Blitbuffer (which is easier to play around with to get the logic right), but I wouldn't know about the Neon optimizations mentioned. Also the Lua version is probably not on par yet with avoiding software rotations, and on the whole @NiLuJe is much better versed in the matter. (I haven't really looked at it in much detail since I did a bunch of stuff for Android/SDL.)

**Comment by rjd22** [comment]
I've tested the color rendering code and verified that is runs but sadly it doesn't seem to work for the PB741. Most likely this is because the screen is different and the code was specially made for the PB Color.

It might for example be that the screen in the PB InkPad Color isn't rotated 90 degree's thus the reverse color rendering doesn't work but I will need to put the screen under the microscope to figure verify that.

**Comment by Frenzie** [comment]
It might be quicker to test the four possible directions in the code? But I'm afraid I can't offer you much assistance.

**Comment by rjd22** [comment]
I've been playing with this but the results keep throwing me on a wild goose chase. I've verified the code is running. I've also checked that it is indeed selecting and sending colors to the `setter` method. This image is shown correctly but it returns it in grayscale.

Might there be something upstream after `setter` that returns the colors to grayscale again? It's really odd that the image shows not color while rendering the grayscale.

For now I'm pausing work on this to get some more experience on kaleido screen rendering.
